examples/linkedlist.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import print_function, absolute_import
+from numba.utils import OrderedDict
+import numpy as np
+from numba import njit
+from numba.jitclass import jitclass
+from numba import int32, deferred_type, optional
+from numba.runtime import rtsys
+
+node_type = deferred_type()
+
+spec = OrderedDict()
+spec['data'] = int32
+spec['next'] = optional(node_type)
+
+
+@jitclass(spec)
+class LinkedNode(object):
+    def __init__(self, data, next):
+        self.data = data
+        self.next = next
+
+    def prepend(self, data):
+        return LinkedNode(data, self)
+
+
+@njit
+def make_linked_node(data):
+    return LinkedNode(data, None)
+
+
+node_type.define(LinkedNode.class_type.instance_type)
```

Comment:
While I do think the deferred_type API can be improved, I do not understand several things from your suggestions:

1) Can you elaborate on "leaking in all aspects of typing and code generation".  Are you referring to the cyclic reference?  

2) I don't understand how `node_type = node_type.resolve(LinkedNode.class_type.instance_type)` would work.  

3) Does `deferred('LinkedNode')` require registering the name "LinkedNode" into the type system?  This seems to make the typesystem more complicated.  There will be name conflict problem to solve as well.